org/mozilla/jss/crypto/KBKDFByteArrayParam.java

```diff
@@ -0,0 +1,51 @@
+package org.mozilla.jss.crypto;
+
+import java.util.Arrays;
+
+import org.mozilla.jss.util.NativeEnclosure;
+
+/**
+ * A KBKDF Byte Array Parameter is a type of KBKDF Data Parameter that
+ * contains a single byte array that gets passed to the KBKDF to be inserted
+ * into the PRF input stream.
+ */
+public class KBKDFByteArrayParam extends KBKDFDataParameter {
+    private byte[] data;
+
+    public KBKDFByteArrayParam() {}
+
+    public KBKDFByteArrayParam(byte[] data) throws IllegalArgumentException {
+        setByteArray(data);
+    }
+
+    public void setByteArray(byte[] data) throws IllegalArgumentException {
+        if (data == null) {
+            throw new IllegalArgumentException("Expected non-null byte array!");
+        }
+
+        this.data = Arrays.copyOf(data, data.length);
+    }
+
+    protected void acquireNativeResources() throws Exception {
```

Comment:
Well, this is to implement the abstract methods from `NativeEnclosure`... The goal of `NativeEnclosure` is to wrap all access to the class at the native level into a single pointer lookup. It's essentially a `Serializable` interface (of sorts), except that it contains a member variables (`ptr` and `ptr_size`) and the requirement that `open(...)` get called before `ptr`/`ptr_size` are accessed, and `close(...)` gets called after all accesses are done. 

Also, you're not going to a real serialization format (XML / Base64 / JSON / ...), you're going to well, a pointer of some type.

So, IMO, having Native in the name 1) points the developer back to `NativeEnclosure` as the source of these methods and 2) is correct and won't become obsolete.

This is also mostly about defining where the boundary is between the Java and the C code is. That boundary can change over time, and could shift so that the abstract class `NativeEnclosure` isn't necessary, but until NSS is rewritten from scratch in Java (and JSS becomes obsolete), you'll still have to interface with C / "native" code somewhere. :-)

Reply:
OK